### PR TITLE
Support negative indices and fix bound checking in symbolic shape inference for Slice

### PR DESCRIPTION
**Pull request by ryotatomioka**
**Description**: 
 * Add `positive=True` for all symbolic variables
 * Add support for negative index

**Motivation and Context**
Currently symbolic shape inference for Slice has the following issues:
1. It does not support bound checking for a symbolic index (e.g., for `x[:L]`, it cannot determine if `L <= len(x)` even if `len(x) = 2 * L`)
2. It does not handle symbolic negative index (e.g., `x[:-L]`)

### CODE REVIEW

**Comment by ryotatomioka** [comment]
> maybe nonnegative? shape could have dim 0 in it.

This is not enough to handle `x[:-L]` if `L` is nonnegative. When does a shape have dim 0?

**Comment by KeDengMS** [comment]
> > maybe nonnegative? shape could have dim 0 in it.
> 
> This is not enough to handle `x[:-L]` if `L` is nonnegative. When does a shape have dim 0?

For example, [NonZero](https://github.com/onnx/onnx/blob/master/docs/Operators.md#NonZero) may have dim 0 if all elements of input are zero.
```
Returns the indices of the elements that are non-zero (in row-major order - by dimension). NonZero behaves similar to numpy.nonzero: https://docs.scipy.org/doc/numpy/reference/generated/numpy.nonzero.html
```

**Comment by ryotatomioka** [comment]
> > > maybe nonnegative? shape could have dim 0 in it.
> 
> > 
> 
> > This is not enough to handle `x[:-L]` if `L` is nonnegative. When does a shape have dim 0?
> 
> 
> 
> For example, [NonZero](https://github.com/onnx/onnx/blob/master/docs/Operators.md#NonZero) may have dim 0 if all elements of input are zero.
> 
> ```
> 
> Returns the indices of the elements that are non-zero (in row-major order - by dimension). NonZero behaves similar to numpy.nonzero: https://docs.scipy.org/doc/numpy/reference/generated/numpy.nonzero.html
> 
> ```

But if I understand correctly, the result of nonzero will not be a named symbolic dimension. Could we assume all **named** dimensions are positive?

**Comment by KeDengMS** [comment]
> > > > maybe nonnegative? shape could have dim 0 in it.
> > 
> > 
> > > 
> > 
> > 
> > > This is not enough to handle `x[:-L]` if `L` is nonnegative. When does a shape have dim 0?
> > 
> > 
> > For example, [NonZero](https://github.com/onnx/onnx/blob/master/docs/Operators.md#NonZero) may have dim 0 if all elements of input are zero.
> > ```
> > 
> > Returns the indices of the elements that are non-zero (in row-major order - by dimension). NonZero behaves similar to numpy.nonzero: https://docs.scipy.org/doc/numpy/reference/generated/numpy.nonzero.html
> > ```
> 
> But if I understand correctly, the result of nonzero will not be a named symbolic dimension. Could we assume all **named** dimensions are positive?

The shape of output of NonZero depends on the data content and has to be [symbolic dimension]()

> > > > maybe nonnegative? shape could have dim 0 in it.
> > 
> > 
> > > 
> > 
> > 
> > > This is not enough to handle `x[:-L]` if `L` is nonnegative. When does a shape have dim 0?
> > 
> > 
> > For example, [NonZero](https://github.com/onnx/onnx/blob/master/docs/Operators.md#NonZero) may have dim 0 if all elements of input are zero.
> > ```
> > 
> > Returns the indices of the elements that are non-zero (in row-major order - by dimension). NonZero behaves similar to numpy.nonzero: https://docs.scipy.org/doc/numpy/reference/generated/numpy.nonzero.html
> > ```
> 
> But if I understand correctly, the result of nonzero will not be a named symbolic dimension. Could we assume all **named** dimensions are positive?



> > > > maybe nonnegative? shape could have dim 0 in it.
> > 
> > 
> > > 
> > 
> > 
> > > This is not enough to handle `x[:-L]` if `L` is nonnegative. When does a shape have dim 0?
> > 
> > 
> > For example, [NonZero](https://github.com/onnx/onnx/blob/master/docs/Operators.md#NonZero) may have dim 0 if all elements of input are zero.
> > ```
> > 
> > Returns the indices of the elements that are non-zero (in row-major order - by dimension). NonZero behaves similar to numpy.nonzero: https://docs.scipy.org/doc/numpy/reference/generated/numpy.nonzero.html
> > ```
> 
> But if I understand correctly, the result of nonzero will not be a named symbolic dimension. Could we assume all **named** dimensions are positive?

The reult of NonZero will be a [symbolic dimension](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/symbolic_shape_infer.py#L809), and it could be 0.

**Comment by ryotatomioka** [comment]
> The reult of NonZero will be a [symbolic dimension](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/symbolic_shape_infer.py#L809), and it could be 0.

I see. Perhaps we can limit the positivity assumption to user named dimension. Does that make sense?

**Comment by KeDengMS** [comment]
> > The reult of NonZero will be a [symbolic dimension](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/symbolic_shape_infer.py#L809), and it could be 0.
> 
> I see. Perhaps we can limit the positivity assumption to user named dimensions. Does that make sense?

Yes that would work. dim 0 should only come from intermediate tensors. Graph inputs should all have positive dims.

**Comment by ryotatomioka** [comment]
I have written the following test and checked that they all pass but it requires `self._use_external_gpu_allocator = False` in `GraphExecutionManager` to run in CPU. Please let me know if whether I should add the test in `onnxruntime_test_python_symbolic_shape_infer.py`.

```python
class TestSymbolicShapeInferenceForSlice(unittest.TestCase):
    def check_dim_value(self, body, expected_dim_value):
        output = unique_element(run_onnx_shape_inference(body, torch.tensor(range(10))))
        dim = unique_element(output.type.tensor_type.shape.dim)
        self.assertEqual(dim.dim_value, expected_dim_value)

    def check_dim_param(self, body, expected_dim_param):
        output = unique_element(run_onnx_shape_inference(body, torch.tensor(range(10))))
        dim = unique_element(output.type.tensor_type.shape.dim)
        self.assertEqual(dim.dim_param, expected_dim_param)

    def test_literal_positive_start_end(self):
        def body(data):
            return data[1:9]
        self.check_dim_param(body, "Min(9, args_0_dim0) - 1")

    def test_literal_negative_start_end(self):
        def body(data):
            return data[-9:-1]
        self.check_dim_value(body, 8)

    def test_negative_start(self):
        def body(data):
            return data[-1:]
        self.check_dim_value(body, 1)

    def test_negative_end(self):
        def body(data):
            return data[:-1]
        self.check_dim_param(body, "args_0_dim0 - 1")

    def test_literal_greater_than_one_end(self):
        def body(data):
            return data[:2]
        self.check_dim_param(body, "Min(2, args_0_dim0)")

    def test_large_end(self):
        def body(data):
            shape = data.shape
            return data[:2 * shape[0]]
        self.check_dim_param(body, "args_0_dim0")

    def test_symbolic_start(self):
        def body(data):
            shape = data.shape
            return data[shape[0]-1:]
        self.check_dim_value(body, 1)

    def test_symbolic_end(self):
        def body(data):
            shape = data.shape
            return data[:shape[0]-1]
        self.check_dim_param(body, "args_0_dim0 - 1")

    def test_literal_start_symbolic_end(self):
        def body(data):
            shape = data.shape
            return data[1:shape[0]-1]
        self.check_dim_param(body, "args_0_dim0 - 2")

    def test_symbolic_negative_half_start(self):
        def body(input1):
            shape = input1.shape
            half = (shape[0] + 1) // 2 # ceil
            return input1[-half:]
        self.check_dim_param(body, "floor(args_0_dim0/2 + 1/2)")

    def test_symbolic_half_end(self):
        def body(input1):
            shape = input1.shape
            half = shape[0] // 2
            return input1[:half]
        self.check_dim_param(body, "floor(args_0_dim0/2)")

    def test_symbolic_negative_half_end(self):
        def body(input1):
            shape = input1.shape
            half = (shape[0] + 1) // 2 # ceil
            return input1[:-half]
        self.check_dim_param(body, "args_0_dim0 - floor(args_0_dim0/2 + 1/2)")


def unique_element(lst):
    assert len(lst) == 1
    return lst[0]


def run_onnx_shape_inference(f, *args, **kwargs):
    class Wrapper(torch.nn.Module):
        def __init__(self):
            super().__init__()
            # we need at least one parameter for get_device_from_module to return
            # a valid device
            self.register_parameter(name='tmp', param=torch.nn.Parameter(torch.randn(1)))

        def forward(self, *args, **kwargs):
            return f(*args, **kwargs)
    model = ORTModule(Wrapper())
    model.eval()
    inference_manager = model._execution_manager._inference_manager
    inference_manager._run_symbolic_shape_infer = True
    print(model(*args, **kwargs))
    return inference_manager._onnx_model.graph.output

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 1]
```diff
@@ -1029,6 +1032,34 @@ def _infer_Size(self, node):
             helper.make_tensor_value_info(node.output[0], onnx.TensorProto.INT64, []))
 
     def _infer_Slice(self, node):
+        def less_equal(x, y):
```
Comment (KeDengMS) [inline onnxruntime/python/tools/symbolic_shape_infer.py:1035]: Would this work? https://docs.sympy.org/latest/modules/core.html?highlight=compare#strictgreaterthan

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 2]
```diff
@@ -1265,8 +1294,8 @@ def _infer_impl(self, start_sympy_data=None):
                 assert s_merge in self.symbolic_dims_
                 self.symbolic_dims_[s] = self.symbolic_dims_[s_merge]
             else:
-                self.symbolic_dims_[s] = sympy.Symbol(s, integer=True)
-
+                # Since inputs are not produced by other ops, we can assume positivity
+                self.symbolic_dims_[s] = sympy.Symbol(s, integer=True, positive=True)
```
Comment (KeDengMS) [inline onnxruntime/python/tools/symbolic_shape_infer.py:1298]: After a second thought, I remember saw inputs with zero dimensions before (i.e. LSTM with zero-length sequence). Can we change this to nonnegative too?

**Comment by KeDengMS** [comment]
> I have written the following test and checked that they all pass but it requires `self._use_external_gpu_allocator = False` in `GraphExecutionManager` to run on CPU. Please let me know if whether I should include the test in `onnxruntime_test_python_symbolic_shape_infer.py`.

Thanks for adding the test. ORTModule currently is only enabled with training, with PyTorch dependency, while `onnxruntime_test_python_symbolic_shape_infer.py` is intended for inference without PyTorch. Would be nice if you can change the test to use [onnx.helper](https://github.com/onnx/onnx/blob/master/docs/PythonAPIOverview.md#creating-an-onnx-model-using-helper-functions) to create the model for symbolic shape inference test.

File: onnxruntime/python/tools/symbolic_shape_infer.py (hunk 2 above)
Comment (ryotatomioka) [reply onnxruntime/python/tools/symbolic_shape_infer.py:1298]: I am afraid not. For example if `L >= 0` (nonnegative), we won't be able to tell if `x[-L:]` has a negative starting index because `L` could be zero. The size of `x[-L:]` is `L` if `L > 0` but `len(x)` if `L == 0`.

Comment (ryotatomioka) [reply onnxruntime/python/tools/symbolic_shape_infer.py:1298]: Do we have a list of ops that can potentially input or output tensors with zero dimensions? Then we can add positivity conditionally based on the producer or consumer of the tensor.

File: onnxruntime/python/tools/symbolic_shape_infer.py (hunk 1 above)
Comment (ryotatomioka) [reply onnxruntime/python/tools/symbolic_shape_infer.py:1035]: I am not completely sure what `StrictGreaterThan` does but I noticed that whether a relational expression can be determined depends on the ordering and some simple reformulation (I believe what the original code did [here](https://github.com/microsoft/onnxruntime/pull/7401/files#diff-b08185b97bd9eefaa587c3f46213db81d7f57a326a52b542f09339070c0f762bL1082) is similar)

As a simple demonstration, the following
```python
from typing import Type
import sympy
from sympy import StrictLessThan, StrictGreaterThan

x = sympy.Symbol("x", integer=True, positive=True)

def try_bool(e):
    try:
        return bool(e)
    except TypeError:
        return f"Cannot determine {e}"

print(try_bool(sympy.floor(x / 2) < x))
print(try_bool(x > sympy.floor(x / 2)))
print(try_bool(StrictLessThan(sympy.floor(x / 2), x)))
print(try_bool(StrictGreaterThan(x, sympy.floor(x / 2))))
```
prints
```
True
Cannot determine x > floor(x/2)
Cannot determine floor(x/2) < x
Cannot determine x > floor(x/2)
```
Generally, sympy seems to have a higher chance of determining boolean expression when the more complicated expression is on the left hand side.

Comment (KeDengMS) [reply onnxruntime/python/tools/symbolic_shape_infer.py:1035]: I see, thanks for the detailed experiments. Please add some comments in the code, as it seems like a limitation or bug in sympy. Besides, maybe worth opening an issue in [sympy](https://github.com/sympy/sympy/issues),

File: onnxruntime/python/tools/symbolic_shape_infer.py (hunk 2 above)
Comment (KeDengMS) [reply onnxruntime/python/tools/symbolic_shape_infer.py:1298]: I see. Seems like a syntax inconsistency in x[-L:] when L >0 and L==0. Let's keep it positive here, and add a comment about this situation of x[-L:] in code.

**Review by KeDengMS — DISMISSED** [review]

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 3]
```diff
@@ -1059,6 +1090,7 @@ def _infer_Slice(self, node):
                     new_sympy_shape[i] = self._new_symbolic_dim_from_output(node, 0, i)
         else:
             for i, s, e, t in zip(axes, starts, ends, steps):
+                e = handle_negative_index(e, new_sympy_shape[i])
```
Comment (KeDengMS) [inline onnxruntime/python/tools/symbolic_shape_infer.py:1093]: >e = handle_negative_index(e, new_sympy_shape[i]) [](http://example.com/codeflow?start=16&length=48)

This seems to cause a test failure in nuphar, with:
(Pdb) new_sympy_shape
[sequence, 1, 10]
(Pdb) starts
[-1]
(Pdb) ends
[-9223372036854775807]
(Pdb) axes
[0]
(Pdb) steps
[-1]
(Pdb) self.int_max_
2147483647

It seems when ends <= -self.int_max_ it should ignore the symbolic dim in handle_negative_index.

File: onnxruntime/python/tools/symbolic_shape_infer.py (hunk 1 above)
Comment (ryotatomioka) [reply onnxruntime/python/tools/symbolic_shape_infer.py:1035]: I see an open issue since 2014. I have added my example as a comment: https://github.com/sympy/sympy/issues/8444

File: onnxruntime/python/tools/symbolic_shape_infer.py (hunk 3 above)
Comment (ryotatomioka) [reply onnxruntime/python/tools/symbolic_shape_infer.py:1093]: Is this `test_loop_to_scan` in `onnxruntime_test_python_nuphar.py`? Is the slice `x[-1::-1]`?

Comment (KeDengMS) [reply onnxruntime/python/tools/symbolic_shape_infer.py:1093]: Yes, it's a negative step in that test case.

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 4]
```diff
@@ -621,7 +624,7 @@ def _infer_Compress(self, node):
                                           output_shape))
 
     def _infer_Concat(self, node):
-        if any([i in self.sympy_data_ for i in node.input]):
+        if any([i in self.sympy_data_ or i in self.initializers_ for i in node.input]):
```
Comment (ryotatomioka) [inline onnxruntime/python/tools/symbolic_shape_infer.py:627]: This was an independent issue I found when I wrote the test. I can make a separate PR if necessary.

File: onnxruntime/test/python/onnxruntime_test_python_symbolic_shape_infer.py [hunk 5]
```diff
@@ -22,5 +27,80 @@ def test_symbolic_shape_infer(self):
...
+        self.assertEqual(shape, ["B", expected_output_dim])
+
+    def test_symbolic_negative_start_index(self):
+        self.check_slice_of_concat(["M", "N"], "-N", "intmax", "one", "N")
+
+    def test_non_unit_step(self):
+        self.check_slice_of_concat(["N", "N"], "zero", "intmax", "two", "N")
+
+    def test_symbolic_step(self):
+        self.check_slice_of_concat(["N", "N"], "zero", "intmax", "N", "floor((3*N - 1)/N)")
```
Comment (ryotatomioka) [inline onnxruntime/test/python/onnxruntime_test_python_symbolic_shape_infer.py:93]: This seems like another limitation of sympy.

**Review by KeDengMS — APPROVED** [review]
Thanks a lot for the fixes and tests!
